File: ruby/word-count/word-count.rb
class Phrase
  def initialize(sentence)
    @sentence = sentence.to_s
  end

  def word_count
    count(words)
  end

  private

  def words
    @sentence.downcase.scan(/\w+/)
  end

  def count(words)
    words.each_with_object(Hash.new(0)) { |word, count| count[word] += 1 }
  end
end

class PhraseWithPrepare
  def initialize(sentence)
    @sentence = prepare(sentence)
  end

  def word_count
    count(words)
  end

  private

  def prepare(sentence)
    sentence.to_s.downcase
  end

  def words
    @sentence.scan(/\w+/)
  end

  def count(words)
    words.each_with_object(Hash.new(0)) { |word, count| count[word] += 1 }
  end
end

File: elixir/word-count/word_count.exs
defmodule Words do
  @doc """
  Count the number of words in the sentence.

  Words are compared case-insensitively.
  """
  @spec count(String.t) :: map()
  def count(sentence) do
    sentence |> String.downcase |> clean_up |> String.split |> group_by_word |> count_grouped |> Enum.into(%{})
  end

  defp clean_up(sentence) do
    String.replace(sentence, ~r/[_:,!?@$%^&]/, " ")
  end

  defp group_by_word(words) do
    Enum.group_by(words, fn(word) -> word end)
  end

  defp count_grouped(words) do
    Enum.map(words, fn({word, list}) -> {word, length(list)} end)
  end
end
